packages/concerto-tools/lib/codegen/fromopenapi/cto/concertometamodelvisitor.js

```diff
@@ -0,0 +1,522 @@
+                    Object.entries(oaiPropertyBody)
+                        .filter(
+                            property => ![
+                                'anyOf', 'oneOf'
+                            ].includes(property[0])
+                        )
+                ),
+                ...(
+                    oaiPropertyBody.anyOf ||
+                    oaiPropertyBody.oneOf
+                )[0]
+            }
+        );
+    }
+    doesPropertyContainAReference({ body: oaiPropertyBody }) {
+        return typeof oaiPropertyBody.$ref === 'string';
+    }
+    resolveReferenceInProperty(
+        {
+            name: oaiPropertyName, body: oaiPropertyBody
+        },
+        parameters
+    ) {
+
+    }
+    isStringLocalReference(oaiReference) {
+        return oaiReference.charAt(0) === '#';
+    }
+    parseLocalReferenceString(oaiReference) {
+        return oaiReference
+            .slice(1, oaiReference.length)
+            .split('/')
+            .filter(str => str.length > 0);
+    }
+    inferPrimitiveConcertoType(
+        {
+            name: oaiPropertyName, body: oaiPropertyBody
+        },
+        parameters
+    ) {
+        if (oaiPropertyBody.type) {
+            switch (oaiPropertyBody.type) {
+            case 'string':
+                if (
+                    isOaiPropertyDateTime(oaiPropertyBody)
+                ) {
+                    return `${parameters.metaModelNamespace}.DateTimeProperty`;
+                }
+
+                if (oaiPropertyBody.format) {
+                    console.warn(
+                        `Format '${oaiPropertyBody.format}' in '${oaiPropertyName}' is not supported. It has been ignored.`
+                    );
+                }
+
+                return `${parameters.metaModelNamespace}.StringProperty`;
+            case 'boolean':
+                return `${parameters.metaModelNamespace}.BooleanProperty`;
+            case 'number':
+                return `${parameters.metaModelNamespace}.DoubleProperty`;
+            case 'integer':
+                return `${parameters.metaModelNamespace}.IntegerProperty`; // Could also be Long?
+            default:
+                throw new Error(
+                    `Type keyword '${oaiPropertyBody.type}' in '${oaiPropertyName}' is not supported`
+                );
+            }
+        }
+    }
+    visitOaiLocalReference({ body: oaiReference }, parameters) {
+        const pathToOaiObject = this.parseLocalReferenceString(oaiReference);
+
+        if (
+            pathToOaiObject[0] === 'components' &&
+            pathToOaiObject[1] === 'schemas'
+        ) {
+            const schemaName = pathToOaiObject[2];
+            return (
+                new OaiSchema(
+                    schemaName,
+                    parameters
+                        .openapidefinition
+                        .components
+                        .schemas[schemaName]
+                )
+            ).accept(this, parameters);
```

Comment:
Can you elaborate on this design, please? I see why you're doing this to honour the visitor pattern, but it feels redundant to both instantiate and then visit the node in the same file. It would make more sense to me if you already had a preprocessing step that gave you a typed AST.

Reply:
Is there an existing JS/TS AST from another library that you could reuse, perhaps?

- https://github.com/apioo/typeschema
- https://www.npmjs.com/package/openapi-typescript

etc..

Reply:
Addressed ✅